File: apps/web/src/utils/converters/formats/light-yaml.ts
/**
 * Light YAML format converter for DiPeO diagrams
 * 
 * Provides a reduced, cleaner YAML format that's more readable than JSON
 * while maintaining all necessary information for diagram reconstruction
 * Works directly with DomainDiagram types
 */

import { stringify, parse } from 'yaml';
import {
  DomainDiagram, DomainNode, DomainArrow, DomainPerson, DomainApiKey, DomainHandle,
  NodeID, ArrowID, PersonID, ApiKeyID, HandleID,
  NodeKind, LLMService, DataType,
  generateNodeId, generateShortId, arrowId, personId, apiKeyId, handleId,
  parseHandleId, createHandleId
} from '@/types';
import { YAML_VERSION, YAML_STRINGIFY_OPTIONS } from '../constants';
import { DomainFormatConverter } from '../core/types';
import { generateNodeHandlesFromRegistry } from '@/utils/node/handle-builder';

// Reduced YAML format types
export interface ReducedYamlNode {
  label: string;
  type: string;
  position: { x: number; y: number };
  // Node-specific data fields
  [key: string]: any;
}

export interface ReducedYamlArrow {
  from: string;      // nodeLabel:handleName
  to: string;        // nodeLabel:handleName
  label?: string;
  [key: string]: any;
}

export interface ReducedYamlPerson {
  model: string;
  service?: string;
  apiKey?: string;   // API key label
  systemPrompt?: string;
  temperature?: number;
  maxTokens?: number;
  topP?: number;
  frequencyPenalty?: number;
  presencePenalty?: number;
}

export interface ReducedYamlApiKey {
  service: string;
}

export interface ReducedYamlHandle {
  node: string;      // Node label
  name: string;      // Handle name
  direction: 'input' | 'output';
  dataType: string;
  position: 'left' | 'right' | 'top' | 'bottom';
  maxConnections?: number;
}

export interface ReducedYamlFormat {
  version: string;
  name?: string;
  description?: string;
  nodes: Record<string, ReducedYamlNode>;
  connections: ReducedYamlArrow[];
  persons?: Record<string, ReducedYamlPerson>;
  apiKeys?: Record<string, ReducedYamlApiKey>;
  handles?: ReducedYamlHandle[];
}

export class LightDomainConverter implements DomainFormatConverter {
  readonly formatName = 'light';
  readonly fileExtension = '.yaml';

  // ID to label mappings
  private nodeIdToLabel = new Map<NodeID, string>();
  private handleIdToLabel = new Map<HandleID, string>();
  private personIdToLabel = new Map<PersonID, string>();
  private apiKeyIdToLabel = new Map<ApiKeyID, string>();
  
  // Label to ID mappings (for deserialization)
  private labelToNodeId = new Map<string, NodeID>();
  private labelToHandleId = new Map<string, HandleID>();
  private labelToPersonId = new Map<string, PersonID>();
  private labelToApiKeyId = new Map<string, ApiKeyID>();

  serialize(diagram: DomainDiagram): string {
    const yamlData = this.toReducedYamlFormat(diagram);
    return stringify(yamlData, YAML_STRINGIFY_OPTIONS);
  }

  deserialize(yamlString: string): DomainDiagram {
    const data = parse(yamlString) as ReducedYamlFormat;
    return this.fromReducedYamlFormat(data);
  }

  private reset(): void {
    this.nodeIdToLabel.clear();
    this.handleIdToLabel.clear();
    this.personIdToLabel.clear();
    this.apiKeyIdToLabel.clear();
    this.labelToNodeId.clear();
    this.labelToHandleId.clear();
    this.labelToPersonId.clear();
    this.labelToApiKeyId.clear();
  }

  /**
   * Convert DomainDiagram to reduced YAML format
   */
  private toReducedYamlFormat(diagram: DomainDiagram): ReducedYamlFormat {
    this.reset();
    
    const nodes: Record<string, ReducedYamlNode> = {};
    const connections: ReducedYamlArrow[] = [];
    const persons: Record<string, ReducedYamlPerson> = {};
    const apiKeys: Record<string, ReducedYamlApiKey> = {};
    const handles: ReducedYamlHandle[] = [];
    
    // Convert API keys first (build ID to label mapping)
    Object.entries(diagram.apiKeys).forEach(([id, key]) => {
      const label = key.label || `API_KEY_${id}`;
      this.apiKeyIdToLabel.set(id as ApiKeyID, label);
      apiKeys[label] = {
        service: key.service
      };
    });
    
    // Convert persons (build ID to label mapping)
    Object.entries(diagram.persons).forEach(([id, person]) => {
      const label = person.label || `PERSON_${id}`;
      this.personIdToLabel.set(id as PersonID, label);
      
      const reducedPerson: ReducedYamlPerson = {
        model: person.model
      };
      
      // Only add non-default values
      if (person.service !== 'openai') reducedPerson.service = person.service;
      if (person.apiKeyId) {
        const apiKeyLabel = this.apiKeyIdToLabel.get(person.apiKeyId);
        if (apiKeyLabel) reducedPerson.apiKey = apiKeyLabel;
      }
      if (person.systemPrompt) reducedPerson.systemPrompt = person.systemPrompt;
      if (person.temperature !== undefined && person.temperature !== 0.2) {
        reducedPerson.temperature = person.temperature;
      }
      if (person.maxTokens) reducedPerson.maxTokens = person.maxTokens;
      if (person.topP !== undefined && person.topP !== 1) reducedPerson.topP = person.topP;
      if (person.frequencyPenalty) reducedPerson.frequencyPenalty = person.frequencyPenalty;
      if (person.presencePenalty) reducedPerson.presencePenalty = person.presencePenalty;
      
      persons[label] = reducedPerson;
    });
    
    // Convert handles (build ID to label mapping)
    Object.entries(diagram.handles).forEach(([id, handle]) => {
      const label = handle.label || `HANDLE_${id}`;
      this.handleIdToLabel.set(id as HandleID, label);
    });
    
    // Convert nodes (build ID to label mapping)
    Object.entries(diagram.nodes).forEach(([id, node]) => {
      const label = String(node.data.label || `NODE_${id}`);
      this.nodeIdToLabel.set(id as NodeID, label);
      
      const reducedNode: ReducedYamlNode = {
        label,
        type: node.type,
        position: {
          x: Math.round(node.position.x),
          y: Math.round(node.position.y)
        }
      };
      
      // Add data fields directly to node (except internal fields)
      Object.entries(node.data).forEach(([key, value]) => {
        // Skip internal/computed fields
        if (key === 'id' || key === 'handles' || key === 'type' || key === 'label') return;
        
        // Replace personId with personLabel
        if (key === 'personId' && value) {
          const personLabel = this.personIdToLabel.get(value as PersonID);
          if (personLabel) {
            reducedNode.person = personLabel;
            return;
          }
        }
        
        // Add other fields directly
        reducedNode[key] = value;
      });
      
      nodes[label] = reducedNode;
    });
    
    // Convert arrows
    Object.entries(diagram.arrows).forEach(([id, arrow]) => {
      // Parse handle IDs to get node IDs and handle labels
      const { nodeId: sourceNodeId, handleLabel: sourceHandleLabel } = parseHandleId(arrow.source);
      const { nodeId: targetNodeId, handleLabel: targetHandleLabel } = parseHandleId(arrow.target);
      
      const sourceNodeLabel = this.nodeIdToLabel.get(sourceNodeId);
      const targetNodeLabel = this.nodeIdToLabel.get(targetNodeId);
      
      if (!sourceNodeLabel || !targetNodeLabel || !sourceHandleLabel || !targetHandleLabel) {
        console.warn(`Skipping arrow ${id} due to missing node or handle labels`);
        return;
      }
      
      const reducedArrow: ReducedYamlArrow = {
        from: `${sourceNodeLabel}:${sourceHandleLabel}`,
        to: `${targetNodeLabel}:${targetHandleLabel}`
      };
      
      // Add data fields if present
      if (arrow.data && Object.keys(arrow.data).length > 0) {
        Object.entries(arrow.data).forEach(([key, value]) => {
          reducedArrow[key] = value;
        });
      }
      
      connections.push(reducedArrow);
    });
    
    // Convert custom handles only
    Object.entries(diagram.handles).forEach(([id, handle]) => {
      const nodeLabel = this.nodeIdToLabel.get(handle.nodeId);
      if (!nodeLabel) return;
      
      // Skip default handles that would be auto-generated
      const node = diagram.nodes[handle.nodeId];
      if (node) {
        const defaultHandles = generateNodeHandlesFromRegistry(node.type as NodeKind, handle.nodeId);
        const isDefault = defaultHandles.some(dh => 
          dh.label === handle.label &&
          dh.direction === handle.direction &&
          dh.dataType === handle.dataType &&
          dh.position === handle.position
        );
        if (isDefault) return;
      }
      
      handles.push({
        node: nodeLabel,
        name: handle.label,
        direction: handle.direction,
        dataType: handle.dataType,
        position: handle.position || 'bottom',
        maxConnections: handle.maxConnections
      });
    });
    
    // Build final format
    const result: ReducedYamlFormat = {
      version: YAML_VERSION,
      nodes,
      connections
    };
    
    // Only add optional fields if they have content
    if (diagram.metadata?.name) result.name = diagram.metadata.name;
    if (diagram.metadata?.description) result.description = diagram.metadata.description;
    if (Object.keys(persons).length > 0) result.persons = persons;
    if (Object.keys(apiKeys).length > 0) result.apiKeys = apiKeys;
    if (handles.length > 0) result.handles = handles;
    
    return result;
  }

  /**
   * Convert from reduced YAML format to DomainDiagram
   */
  private fromReducedYamlFormat(data: ReducedYamlFormat): DomainDiagram {
    this.reset();
    
    const diagram: DomainDiagram = {
      nodes: {},
      arrows: {},
      handles: {},
      persons: {},
      apiKeys: {}
    };
    
    // Import API keys first
    if (data.apiKeys) {
      Object.entries(data.apiKeys).forEach(([label, apiKey]) => {
        const id = apiKeyId(generateShortId());
        this.labelToApiKeyId.set(label, id);
        diagram.apiKeys[id] = {
          id,
          label,
          service: apiKey.service as LLMService
        };
      });
    }
    
    // Import persons
    if (data.persons) {
      Object.entries(data.persons).forEach(([label, person]) => {
        const id = personId(generateShortId());
        this.labelToPersonId.set(label, id);
        
        let apiKeyId: ApiKeyID | undefined;
        if (person.apiKey) {
          apiKeyId = this.labelToApiKeyId.get(person.apiKey);
        }
        
        diagram.persons[id] = {
          id,
          label,
          model: person.model,
          service: (person.service || 'openai') as LLMService,
          apiKeyId,
          systemPrompt: person.systemPrompt || '',
          temperature: person.temperature ?? 0.2,
          maxTokens: person.maxTokens,
          topP: person.topP ?? 1,
          frequencyPenalty: person.frequencyPenalty ?? 0,
          presencePenalty: person.presencePenalty ?? 0
        };
      });
    }
    
    // Import nodes
    Object.entries(data.nodes).forEach(([label, node]) => {
      const nodeId = generateNodeId();
      this.labelToNodeId.set(label, nodeId);
      
      // Extract node properties
      const { type, position, label: nodeLabel, ...restData } = node;
      
      const nodeData: any = {
        label,
        ...restData
      };
      
      // Convert person reference
      if (nodeData.person) {
        const personId = this.labelToPersonId.get(nodeData.person);
        if (personId) {
          nodeData.personId = personId;
          delete nodeData.person;
        }
      }
      
      diagram.nodes[nodeId] = {
        id: nodeId,
        type: type as NodeKind,
        position,
        data: nodeData
      };
    });
    
    // Generate default handles for nodes
    Object.values(diagram.nodes).forEach(node => {
      const handles = generateNodeHandlesFromRegistry(node.type as NodeKind, node.id);
      handles.forEach(handle => {
        this.labelToHandleId.set(handle.label, handle.id);
        diagram.handles[handle.id] = handle;
      });
    });
    
    // Import custom handles
    if (data.handles) {
      data.handles.forEach(handle => {
        const nodeId = this.labelToNodeId.get(handle.node);
        if (!nodeId) return;
        
        const id = createHandleId(nodeId, handle.name);
        this.labelToHandleId.set(handle.name, id);
        
        diagram.handles[id] = {
          id,
          nodeId,
          label: handle.name,
          direction: handle.direction,
          dataType: handle.dataType as DataType,
          position: handle.position,
          maxConnections: handle.maxConnections
        };
      });
    }
    
    // Import connections
    data.connections.forEach(connection => {
      const [fromNode, fromHandle] = connection.from.split(':');
      const [toNode, toHandle] = connection.to.split(':');
      
      const sourceNodeId = this.labelToNodeId.get(fromNode || '');
      const targetNodeId = this.labelToNodeId.get(toNode || '');
      
      if (!sourceNodeId || !targetNodeId) {
        console.warn(`Skipping connection ${connection.from} -> ${connection.to} due to missing nodes`);
        return;
      }
      
      // Create handle IDs from node IDs and handle names
      const sourceHandleId = createHandleId(sourceNodeId, fromHandle || 'output');
      const targetHandleId = createHandleId(targetNodeId, toHandle || 'input');
      
      const id = arrowId(generateShortId());
      const { from, to, ...arrowData } = connection;
      
      diagram.arrows[id] = {
        id,
        source: sourceHandleId,
        target: targetHandleId,
        data: arrowData
      };
    });
    
    // Add metadata
    if (data.name || data.description) {
      diagram.metadata = {
        id: generateShortId(),
        name: data.name,
        description: data.description,
        version: data.version || '2.0.0',
        created: new Date().toISOString(),
        modified: new Date().toISOString()
      };
    }
    
    return diagram;
  }
}

// Export convenience functions
export const toLightYAML = (diagram: DomainDiagram): string => {
  const converter = new LightDomainConverter();
  return converter.serialize(diagram);
};

export const fromLightYAML = (yamlString: string): DomainDiagram => {
  const converter = new LightDomainConverter();
  return converter.deserialize(yamlString);
};

File: server/src/domains/diagram/converters/light_yaml.py
"""Light YAML format converter - refactored to use enhanced base and shared components."""
import uuid
from typing import Dict, Any, List, Optional

from .base import YamlBasedConverter
from ..models.domain import (
    DomainDiagram, DomainNode, DomainArrow, DomainPerson, 
    DomainApiKey, DiagramMetadata, Vec2,
    NodeType, LLMService, ForgettingMode
)


class LightYamlConverter(YamlBasedConverter):
    """Converts between DomainDiagram and light YAML format using shared components."""
    
    def __init__(self):
        super().__init__()
        self.label_to_node_id = {}
        self.node_id_to_label = {}
        self.person_label_to_id = {}
        self.person_id_to_label = {}
    
    # Implement abstract methods from EnhancedDiagramConverter
    def extract_nodes(self, data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Extract node data from light YAML format."""
        return data.get('nodes', [])
    
    def extract_arrows(self, data: Dict[str, Any], diagram: DomainDiagram) -> List[Dict[str, Any]]:
        """Extract edge data from light YAML format connections."""
        edges = []
        for conn in data.get('connections', []):
            source_id = self.label_to_node_id.get(conn['from'])
            target_id = self.label_to_node_id.get(conn['to'])
            
            if source_id and target_id:
                # Use default handles
                source_handle = f"{source_id}:output"
                target_handle = f"{target_id}:input"
                
                edges.append({
                    'id': f"arrow_{uuid.uuid4().hex[:8]}",
                    'source': source_handle,
                    'target': target_handle,
                    'data': conn.get('data', {})
                })
        
        return edges
    
    def extract_node_id(self, node_data: Dict[str, Any], index: int) -> str:
        """Generate node ID and maintain label mapping."""
        node_id = f"{node_data.get('type', 'node')}_{uuid.uuid4().hex[:8]}"
        label = node_data.get('label', f"Node_{index}")
        
        # Maintain bidirectional mapping
        self.label_to_node_id[label] = node_id
        self.node_id_to_label[node_id] = label
        
        return node_id
    
    def extract_node_properties(self, node_data: Dict[str, Any]) -> Dict[str, Any]:
        """Extract node properties including label and person assignment."""
        properties = {'label': node_data.get('label', '')}
        
        # Assign person if specified
        if 'person' in node_data and node_data['person'] in self.person_label_to_id:
            properties['personId'] = self.person_label_to_id[node_data['person']]
        
        # Add props
        if 'props' in node_data:
            properties.update(node_data['props'])
        
        return properties
    
    def post_process_diagram(self, diagram: DomainDiagram, data: Dict[str, Any]) -> None:
        """Process persons and metadata after nodes and edges."""
        # Clear mappings for next use
        self.label_to_node_id.clear()
        self.node_id_to_label.clear()
        
        # Convert persons
        self._process_persons(diagram, data)
        
        # Create metadata
        diagram.metadata = DiagramMetadata(
            name=data.get('name', 'Imported Diagram'),
            description=data.get('description', ''),
            version='2.0.0'
        )
        
        # Convert arrows from edges
        diagram.arrows = {}
        for edge in diagram.edges.values():
            arrow = DomainArrow(
                id=edge.id,
                source=edge.source,
                target=edge.target,
                data=edge.data if hasattr(edge, 'data') else {}
            )
            diagram.arrows[arrow.id] = arrow
        
        # Clear edges as light format uses arrows
        diagram.edges = {}
    
    def _process_persons(self, diagram: DomainDiagram, data: Dict[str, Any]) -> None:
        """Process persons from light YAML data."""
        diagram.persons = {}
        diagram.api_keys = {}
        
        for person_data in data.get('persons', []):
            person_id = f"person_{uuid.uuid4().hex[:8]}"
            label = person_data['label']
            self.person_label_to_id[label] = person_id
            self.person_id_to_label[person_id] = label
            
            # Create a default API key for the person
            api_key_id = f"apikey_{uuid.uuid4().hex[:8]}"
            diagram.api_keys[api_key_id] = DomainApiKey(
                id=api_key_id,
                label=f"{label}_key",
                service=LLMService(person_data['service']),
                key="YOUR_API_KEY_HERE"  # Placeholder
            )
            
            diagram.persons[person_id] = DomainPerson(
                id=person_id,
                label=label,
                service=LLMService(person_data['service']),
                model=person_data['model'],
                systemPrompt=person_data.get('systemPrompt'),
                api_key_id=api_key_id,
                forgettingMode=ForgettingMode(person_data.get('forgettingMode', 'none'))
            )
    
    def _calculate_format_confidence(self, data: Dict[str, Any]) -> float:
        """Calculate confidence that this is light YAML format."""
        confidence = 0.0
        
        # Check for version marker
        if data.get('version') == 'light':
            confidence += 0.5
        
        # Check for expected structure
        if 'nodes' in data and isinstance(data['nodes'], list):
            confidence += 0.2
            # Check if nodes use labels
            if data['nodes'] and all('label' in node for node in data['nodes'][:3]):
                confidence += 0.2
        
        # Check for connections instead of edges
        if 'connections' in data and isinstance(data['connections'], list):
            confidence += 0.1
        
        return min(confidence, 1.0)
    
    # Serialization methods
    def diagram_to_data(self, diagram: DomainDiagram) -> Dict[str, Any]:
        """Convert diagram to light YAML data structure."""
        # Create label mappings
        node_labels = {node_id: node.data.get('label', f'Node_{idx}') 
                      for idx, (node_id, node) in enumerate(diagram.nodes.items())}
        person_labels = {person_id: person.label 
                        for person_id, person in diagram.persons.items()}
        
        # Convert nodes with labels
        nodes = []
        for node_id, node in diagram.nodes.items():
            light_node = {
                'label': node_labels[node_id],
                'type': node.type if isinstance(node.type, str) else node.type.value,
                'position': {'x': node.position.x, 'y': node.position.y}
            }
            
            # Add person by label if assigned
            person_id = node.data.get('personId')
            if person_id and person_id in person_labels:
                light_node['person'] = person_labels[person_id]
            
            # Add props if present
            props = {k: v for k, v in node.data.items() 
                    if k not in ['label', 'personId']}
            if props:
                light_node['props'] = props
            
            nodes.append(light_node)
        
        # Convert arrows with label references
        connections = []
        arrows = diagram.arrows if diagram.arrows else diagram.edges
        for arrow in arrows.values():
            # Extract node IDs from handle IDs (format: "nodeId:handleName")
            source_node_id = arrow.source.split(':')[0]
            target_node_id = arrow.target.split(':')[0]
            
            source_label = node_labels.get(source_node_id, source_node_id)
            target_label = node_labels.get(target_node_id, target_node_id)
            
            light_arrow = {
                'from': source_label,
                'to': target_label
            }
            
            arrow_data = arrow.data if hasattr(arrow, 'data') else {}
            if arrow_data:
                light_arrow['data'] = arrow_data
            
            connections.append(light_arrow)
        
        # Convert persons
        persons = []
        for person in diagram.persons.values():
            light_person = {
                'label': person.label,
                'service': person.service.value,
                'model': person.model
            }
            
            if person.systemPrompt:
                light_person['systemPrompt'] = person.systemPrompt
            
            if person.forgettingMode != ForgettingMode.NONE:
                light_person['forgettingMode'] = person.forgettingMode.value
            
            persons.append(light_person)
        
        # Build light format
        data = {
            'version': 'light',
            'nodes': nodes,
            'connections': connections
        }
        
        if persons:
            data['persons'] = persons
        
        # Add metadata if present
        if diagram.metadata and diagram.metadata.name:
            data['name'] = diagram.metadata.name
        if diagram.metadata and diagram.metadata.description:
            data['description'] = diagram.metadata.description
        
        return data